src/adapter/src/session.rs

```diff
@@ -1457,3 +1466,181 @@ impl From<&ExplainContext> for RequireLinearization {
+    pub fn builder(sources: impl IntoIterator<Item = GlobalId>) -> WriteLocksBuilder {
+        let locks = sources.into_iter().map(|gid| (gid, None)).collect();
+        WriteLocksBuilder { locks }
+    }
+
+    /// Validate this set of [`WriteLocks`] is sufficient for the provided collections.Dropping the
+    /// currently held locks if it's not.
+    pub fn validate(
+        self,
+        collections: impl Iterator<Item = GlobalId>,
+    ) -> Result<Self, BTreeSet<GlobalId>> {
+        let mut missing = BTreeSet::new();
+        for collection in collections {
+            if !self.locks.contains_key(&collection) {
+                missing.insert(collection);
+            }
+        }
+
+        if missing.is_empty() {
+            Ok(self)
+        } else {
+            // Explicitly drop the already acquired locks.
+            drop(self);
+            Err(missing)
+        }
+    }
+}
+
+impl Drop for WriteLocks {
+    fn drop(&mut self) {
+        // We may have merged the locks into GroupCommitWriteLocks, thus it could be empty.
+        if !self.locks.is_empty() {
+            tracing::info!(
+                conn_id = %self.conn_id,
+                locks = ?self.locks,
+                "dropping write locks",
+            );
+        }
+    }
+}
+
+/// A builder struct that helps us acquire all of the locks we need, or none of them.
+///
+/// See [`WriteLocks::builder`].
+#[derive(Debug)]
+pub struct WriteLocksBuilder {
+    locks: BTreeMap<GlobalId, Option<tokio::sync::OwnedMutexGuard<()>>>,
+}
+
+impl WriteLocksBuilder {
+    /// Adds a lock to this builder.
+    pub fn insert_lock(&mut self, id: GlobalId, lock: tokio::sync::OwnedMutexGuard<()>) {
+        self.locks.insert(id, Some(lock));
+    }
+
+    /// Finish this builder by returning either all of the necessary locks, or none of them.
+    pub fn all_or_nothing(self, conn_id: &ConnectionId) -> Result<WriteLocks, BTreeSet<GlobalId>> {
+        let (locks, missing): (BTreeMap<_, _>, BTreeSet<_>) =
+            self.locks
+                .into_iter()
+                .partition_map(|(gid, lock)| match lock {
+                    Some(lock) => itertools::Either::Left((gid, lock)),
+                    None => itertools::Either::Right(gid),
+                });
+
+        if missing.is_empty() {
+            tracing::info!(%conn_id, ?locks, "acquired write locks");
+            Ok(WriteLocks {
+                locks,
+                conn_id: conn_id.clone(),
+            })
+        } else {
+            // Explicitly drop the already acquired locks.
+            drop(locks);
+            Err(missing)
+        }
+    }
+}
+
+/// Collection of [`WriteLocks`] gathered during [`group_commit`].
+///
+/// Note: This struct should __never__ be used outside of group commit because it attempts to merge
+/// together several collections of [`WriteLocks`] which if not done carefully can cause deadlocks
+/// or consistency violations.
```

Comment:
Should we move it to `append.rs` then?

Reply:
I tried to do that originally, but then we'd need to make the inner `locks` field on `WriteLocks` public, or provide a drain-like method on `WriteLocks`, both of which I wanted to avoid.